### PR TITLE
Add reset subcommand for sidebar

### LINKED ISSUE #3063: Introduce reset subcommand for sidebar
Proposal:

`wp sidebar reset <sidebar-id>`

- Cleans up all widgets in the the given sidebar.
- If sidebar ID is not passed, all registered sidebars will be reset.

Technically, reset means moving widgets from sidebar to `Inactive Widgets`.

### PR DESCRIPTION
**Pull request by ernilambar**
Fixes #3063

### CODE REVIEW

**Comment by danielbachhuber** [comment]
For this command, can you create a separate `features/sidebar-reset.feature` file and move your tests in there?

File: features/sidebar-reset.feature [hunk 1]
```diff
@@ -0,0 +1,67 @@
+Feature: Manage WordPress sidebars
```
Comment (danielbachhuber) [inline features/sidebar-reset.feature]: This should probably be "Reset WordPress sidebars"

File: features/sidebar-reset.feature [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    Given a WP install
+
+    When I run `wp theme install p2 --activate`
+    Then STDOUT should not be empty
+
+    When I run `wp widget list sidebar-1 --format=count`
+    Then STDOUT should be:
+    """
+    6
+    """
```
Comment (danielbachhuber) [inline features/sidebar-reset.feature]: The indentation is slightly off here. It should be:

```
Then STDOUT should be:
  """
  6
  """
```

This comment applies to all instances throughout.

File: php/commands/sidebar.php [hunk 3]
```diff
@@ -73,6 +73,204 @@ public function list_( $args, $assoc_args ) {
...
+	 *
+	 * ## OPTIONS
+	 *
+	 * [<sidebar-id>]
+	 * : ID for the corresponding sidebar.
+	 *
+	 * @subcommand reset
+	 */
+	public function reset( $args, $assoc_args ) {
+		if ( empty( $args ) ) {
```
Comment (danielbachhuber) [inline php/commands/sidebar.php]: We should follow the existing pattern of requiring one or more sidebars to be passed or an `--all` flag.

File: php/commands/sidebar.php [hunk 4]
```diff
@@ -73,6 +73,204 @@ public function list_( $args, $assoc_args ) {
 		$formatter->display_items( $wp_registered_sidebars );
 	}
 
+	/**
+	 * Reset one or all sidebars.
+	 *
```
Comment (danielbachhuber) [inline php/commands/sidebar.php]: On a new line, could you explain what "resetting" a sidebar means?

File: php/commands/sidebar.php [hunk 5]
```diff
@@ -73,6 +73,204 @@ public function list_( $args, $assoc_args ) {
...
+	 * ## OPTIONS
+	 *
+	 * [<sidebar-id>]
+	 * : ID for the corresponding sidebar.
+	 *
+	 * @subcommand reset
```
Comment (danielbachhuber) [inline php/commands/sidebar.php]: This definition isn't needed if the command name is the same as the method name.

File: php/commands/sidebar.php (hunk 4 above)
Comment (danielbachhuber) [inline php/commands/sidebar.php]: Also, can you include one or more examples in the new pattern we've established?

File: features/sidebar-reset.feature [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    When I run `wp sidebar reset sidebar-1`
+    And I run `wp widget list sidebar-1 --format=count`
+    Then STDOUT should be:
+    """
+    0
+    """
```
Comment (danielbachhuber) [inline features/sidebar-reset.feature]: Should there be a test for where the widgets are expected to end up?

File: php/commands/sidebar.php [hunk 7]
```diff
@@ -73,6 +73,204 @@ public function list_( $args, $assoc_args ) {
...
+			WP_CLI::success( sprintf( "Sidebar '%s' reset.", $sidebar_id ) );
+		}
+	}
+
+	/**
+	 * Check whether a sidebar is a valid sidebar.
+	 *
+	 * @param string $sidebar_id Sidebar ID.
+	 */
+	private function validate_sidebar( $sidebar_id ) {
```
Comment (danielbachhuber) [inline php/commands/sidebar.php]: Given this method is only used once, can't we just include its logic in the `reset()` method?

**Comment by danielbachhuber** [comment]
@ernilambar I just realized this uses a lot of code copied from `class Widget_Command`. Could we make this `wp widget reset` or similar, as to reuse the code?

**Comment by ernilambar** [comment]
I also felt that when I had to copy functions from Widget command. I even had this crazy idea of using launch_self for some tasks, but on searching I could only found two places where it is used now. So, I assumed it is best to avoid in core commands.

Since this is basically related to sidebar, I tried to keep it under sidebar command. Adding it under widget is also one approach I guess.

`wp widget reset`
Are we up for this?

1. `wp widget reset sidebar-1`
2. `wp widget reset sidebar-1 sidebar-2`
3. `wp widget reset --all`
4. `wp widget reset` [This should also throw error I believe. Like we do in plugin update - `Error: Please specify one or more plugins, or use --all.` ]
5. `wp widget reset wp_inactive_widgets` [Should throw error.]

Re 5: There is button `Clean Inactive Widgets` in the `/wp-admin/widgets.php`. This functionality could be incorporated for command in 5. Or may be we should implement separately creating separate ticket?

**Pull request closed without merging**